upcoming: [DI-19819] - Added a filter builder component for ACLP that builds the filters per service type

venkymano opened the pull request:
## Description 📝
The CloudPulseDashboardFilterBuilder component builds the required mandatory global filters from a JSON config based on service types like linode, dbass etc and the CloudPulseDashboard Landing component pushes the selected filters to the dashboard component

## Changes  🔄
List any change relevant to the reviewer.
- Added a new CloudPulseDashboardFilterBuilder that will build the global filters based on FilterConfig.ts per service type
- Added a new CloudPulseComponentRenderer component that builds the react element based on filterKey
- Added UT's for both of the component
- Complete integration of the CloudPulseDashboardFilterBuilder in the GlobalFilters.tsx
- Added logic in the CloudPulseDashboardLanding component to pass the filters selected to the dashboard component
- Changes for moving CloudPulseTimeRangeSelect from **select to Autocomplete**
- CSS changes making the filters compatible with different screens like desktop, tablet and mobile.
- The filter builder component, shows a scroll bar after two rows
- Added a new monitor icon svg as per UX mockups

## Target release date 🗓️
13th August 2024

## Preview 📷
**Include a screenshot or screen recording of the change**

:bulb: Use `<video src="" />` tag when including recordings in table.

| Before  | After   |
| ------- | ------- |
|![Screenshot 2024-07-26 at 3 22 24 PM](https://github.com/user-attachments/assets/38259036-72d7-4296-8ab1-3f15fe28090e)|![Screenshot 2024-08-07 at 6 18 53 PM](https://github.com/user-attachments/assets/bd3f4495-fbe5-4a2c-85a3-32b947fdd21a)|
|![Screenshot 2024-08-07 at 6 20 40 PM](https://github.com/user-attachments/assets/c914a79f-fa6e-49fa-a71a-189b604520dc)|![Screenshot 2024-08-07 at 6 20 52 PM](https://github.com/user-attachments/assets/8a26a55e-0972-4bcc-a9f0-d744326c2dfc)|
|![Screenshot 2024-08-07 at 6 21 29 PM](https://github.com/user-attachments/assets/f666587b-fa2e-456e-b91a-d57683507cfd)|![Screenshot 2024-08-07 at 6 21 39 PM](https://github.com/user-attachments/assets/8375b846-ecfa-44a4-8ce6-0b8a8827f87d)|
|![Screenshot 2024-08-07 at 6 22 11 PM](https://github.com/user-attachments/assets/181c1bb2-35a4-4a6e-b819-f626b8491863)|![Screenshot 2024-08-07 at 6 22 33 PM](https://github.com/user-attachments/assets/519a1001-3b5a-47ab-b195-3bddf1cf3493)|
|![Screenshot 2024-08-07 at 6 40 47 PM](https://github.com/user-attachments/assets/7d32233a-e885-4d32-8f2d-a7da3117653a)|![Screenshot 2024-08-07 at 6 42 59 PM](https://github.com/user-attachments/assets/cf15804c-e24e-4586-a5c7-cf2ea40e70a1)|


## How to test 🧪

1. Enable mock service as some endpoints are not ready to use
2. Go the Monitor Landing Page
4. Select a Dashboard from dashboard selection dropdown
5. You will be able to see the filters build below the existing global filters
6. After selecting all the mandatory filters, you will able to see the widgets being rendered

## As an Author I have considered 🤔

*Check all that apply*

- [x] 👀 Doing a self review
- [x] ❔ Our [contribution guidelines](https://github.com/linode/manager/blob/develop/docs/CONTRIBUTING.md)
- [x] 🤏 Splitting feature into small PRs
- [x] ➕ Adding a changeset
- [x] 🧪 Providing/Improving test coverage
- [x] 🔐 Removing all sensitive information from the code and PR description
- [x] 🚩 Using a feature flag to protect the release
- [ ] 👣 Providing comprehensive reproduction steps
- [ ] 📑 Providing or updating our documentation
- [ ] 🕛 Scheduling a pair reviewing session
- [x] 📱 Providing mobile support
- [ ] ♿  Providing accessibility support

nikhagra reviewed (approved):
looks good

jaalah-akamai commented on packages/manager/src/features/CloudPulse/Utils/FilterBuilder.ts:
```diff
@@ -0,0 +1,103 @@
+    const serviceTypeConfig = FILTER_CONFIG.get(dashboard.service_type!);
+    const filters = serviceTypeConfig?.filters ?? [];
+
+    for (const filter of filters) {
```
Nit: We could potentially save some iterations here by using `.find`:
```
const filter = filters.find(
  (filter) =>
    filter?.configuration.filterKey === filterKey &&
    filter.configuration.dependency
);

if (filter) {
  // ... other stuff
}
```

venkymano replied:
done

jaalah-akamai commented on packages/manager/src/features/CloudPulse/shared/CloudPulseDashboardFilterBuilder.tsx:
```diff
@@ -0,0 +1,255 @@
+
+export interface CloudPulseDashboardFilterBuilderProps {
+  /**
+   * we need to dashboard here, as we can infer serviceType and other required properties from it.
```
```suggestion
   * We need the dashboard here, as we can infer serviceType and other required properties from it.
```

venkymano replied:
done

jaalah-akamai commented on packages/manager/src/features/CloudPulse/shared/CloudPulseDashboardFilterBuilder.tsx:
```diff
@@ -0,0 +1,255 @@
+      };
+    };
+
+    const getProps = (config: CloudPulseServiceTypeFilters) => {
```
Do we think this should be wrapped in a `useCallback`?

venkymano replied:
Thanks, I have wrapped this  method as well most of the methods in this component around callback now, and everything works fine

jaalah-akamai commented on packages/manager/src/features/CloudPulse/shared/CloudPulseDashboardFilterBuilder.tsx:
```diff
@@ -0,0 +1,255 @@
+            <Typography>Filters</Typography>
+          </Button>
+        )}
+        <Grid container xs={12}>
+          {showFilter &&
+            FILTER_CONFIG.get(props.dashboard.service_type)
+              ?.filters.filter(
+                (config) =>
+                  (props.isServiceAnalyticsIntegration === false &&
+                    config.configuration.neededInServicePage === false) ||
+                  (props.isServiceAnalyticsIntegration === true &&
+                    config.configuration.neededInServicePage === true)
+              )
+              .map((filter, index) => {
+                return (
+                  <Grid
+                    key={filter.configuration.filterKey}
+                    sx={{ marginLeft: 2 }}
+                    xs
+                  >
+                    {RenderComponent({
+                      ...getProps(filter),
+                      key: index + filter.configuration.filterKey,
+                    })}
+                  </Grid>
+                );
+              })}
+        </Grid>
```
We should consider wrapping the <Grid> component with the showFilter condition to handle cases where FILTER_CONFIG.get is undefined. Additionally, I recommend moving this logic outside of the JSX for better readability. It might also be beneficial to memoize these computations, as there are several seemingly expensive operations in this file.

venkymano replied:
I have moved the part where we render the component after getting filters into a new function and handled the undefined error case of filters as well

jaalah-akamai commented on packages/manager/src/features/CloudPulse/shared/CloudPulseDashboardFilterBuilder.tsx:
```diff
@@ -0,0 +1,255 @@
+  newProps: CloudPulseDashboardFilterBuilderProps
+) {
+  return (
+    oldProps.dashboard &&
+    newProps.dashboard &&
+    oldProps.dashboard.id == newProps.dashboard.id &&
+    !newProps.isServiceAnalyticsIntegration
```
```suggestion
    oldProps.dashboard?.id === newProps.dashboard?.id &&
    !newProps.isServiceAnalyticsIntegration
```

venkymano replied:
done

jaalah-akamai commented on packages/manager/src/features/CloudPulse/shared/CloudPulseResourcesSelect.tsx:
```diff
@@ -17,19 +17,25 @@ export interface CloudPulseResources {
     const { data: resources, isLoading } = useResourcesQuery(
-      props.region && props.resourceType ? true : false,
+      props.disabled != undefined
+        ? !props.disabled
+        : props.region && props.resourceType
+        ? true
+        : false,
```
```suggestion
        ? !props.disabled 
        : Boolean(props.region && props.resourceType);
```

venkymano replied:
done

jaalah-akamai commented on packages/manager/src/features/CloudPulse/shared/CloudPulseDashboardFilterBuilder.tsx:
```diff
@@ -0,0 +1,255 @@
+
+    return (
+      <>
+        {!showFilter && (
+          <Button onClick={toggleShowFilter} sx={{ marginTop: 2 }}>
+            <KeyboardArrowRightIcon />
+            <Typography>Filters</Typography>
+          </Button>
+        )}
+        {showFilter && (
+          <Button onClick={toggleShowFilter} sx={{ marginTop: 2 }}>
+            <KeyboardArrowDownIcon />
+            <Typography>Filters</Typography>
+          </Button>
+        )}
```
```suggestion
        <Button onClick={toggleShowFilter} sx={{ marginTop: 2 }}>
          {showFilter ? <KeyboardArrowDownIcon /> : <KeyboardArrowRightIcon />}
          <Typography>Filters</Typography>
        </Button>
```

venkymano replied:
done

jaalah-akamai commented on packages/manager/src/features/CloudPulse/shared/CloudPulseDashboardFilterBuilder.tsx:
```diff
@@ -0,0 +1,255 @@
+            FILTER_CONFIG.get(props.dashboard.service_type)
+              ?.filters.filter(
+                (config) =>
+                  (props.isServiceAnalyticsIntegration === false &&
+                    config.configuration.neededInServicePage === false) ||
+                  (props.isServiceAnalyticsIntegration === true &&
+                    config.configuration.neededInServicePage === true)
```
```suggestion
                  (!isServiceAnalyticsIntegration && !config.configuration.neededInServicePage) ||
                  (isServiceAnalyticsIntegration && config.configuration.neededInServicePage)
```

jaalah-akamai commented on packages/manager/src/features/CloudPulse/shared/CloudPulseDashboardFilterBuilder.tsx:
```diff
@@ -0,0 +1,255 @@
+
+export const CloudPulseDashboardFilterBuilder = React.memo(
+  (props: CloudPulseDashboardFilterBuilderProps) => {
+    const [, setDependentFilters] = React.useState<{
```
Nit: Our convention is to destructure the props `const { ... } = props;`

venkymano replied:
destructured and used

jaalah-akamai reviewed (commented):
Looks good overall, just some cleanup 👍

bnussman-akamai commented on packages/manager/src/features/CloudPulse/shared/CloudPulseResourcesSelect.tsx:
```diff
@@ -17,19 +17,23 @@ export interface CloudPulseResources {
   placeholder?: string;
   region: string | undefined;
   resourceType: string | undefined;
+  xFilter?: { [key: string]: any } | undefined;
```
Seeing a ton of `any`s in this PR. If this is going to use standard API filter, we have a type for API filtering called `Filter`.

venkymano replied:
Hi @bnussman-akamai , I removed most of the any's in the PR now as I could, I also build a logic to build and pass xFilter here as Filter object as per your suggestion

jaalah-akamai commented:
@bnussman-akamai and @carrillo-erik  this should be good for another pass 👍

bnussman-akamai commented on packages/manager/src/features/CloudPulse/shared/CloudPulseComponentRenderer.test.tsx:
```diff
@@ -0,0 +1,79 @@
+    };
+
+    const { getByPlaceholderText } = renderWithTheme(
+      <Grid item key={'region'} sx={{ marginLeft: 2 }} xs>
```
Why just a random `key`?

VenkatMano replied:
removed, we don't need

venkymano replied:
removed

bnussman-akamai commented on packages/manager/src/features/CloudPulse/shared/CloudPulseDashboardFilterBuilder.tsx:
```diff
@@ -0,0 +1,263 @@
+      !FILTER_CONFIG.has(dashboard.service_type)
+    ) {
+      return (
+        <ErrorState
+          CustomIcon={InfoIcon}
+          CustomIconStyles={{ height: '10%', width: '10%' }}
+          errorText={'Please pass valid dashboard to render the filters'}
+        ></ErrorState>
```
If there are no children, can we use a self-closing tag?

```suggestion
        <ErrorState
          CustomIcon={InfoIcon}
          CustomIconStyles={{ height: '10%', width: '10%' }}
          errorText={'Please pass valid dashboard to render the filters'}
        />
```

venkymano replied:
done

venkymano replied:
done

bnussman-akamai commented on packages/manager/src/features/CloudPulse/Utils/models.ts:
Would be nice to see some jsdoc style comments explaining what there types and properties mean

bnussman-akamai commented on packages/manager/src/features/CloudPulse/Overview/GlobalFilters.tsx:
```diff
@@ -122,6 +124,11 @@ export const GlobalFilters = React.memo((props: GlobalFilterProperties) => {
           </Tooltip>
         </Grid>
       </StyledGrid>
+      <CloudPulseDashboardFilterBuilder // this simple integration is for test purpose, the functionality integration will be done in upcoming PR by driving all filters through filterbuilder
+        dashboard={selectedDashboard!}
+        emitFilterChange={() => {}}
+        isServiceAnalyticsIntegration={false}
+      ></CloudPulseDashboardFilterBuilder>
```
Can we do this because there are no children?

```suggestion
      <CloudPulseDashboardFilterBuilder // this simple integration is for test purpose, the functionality integration will be done in upcoming PR by driving all filters through filterbuilder
        dashboard={selectedDashboard!}
        emitFilterChange={() => {}}
        isServiceAnalyticsIntegration={false}
      />
```

VenkatMano replied:
done

venkymano replied:
done

venkymano replied:
done

jdamore-linode commented on packages/manager/src/features/CloudPulse/Utils/FilterConfig.ts:
```diff
@@ -0,0 +1,130 @@
+      name: 'Time Duration',
+    },
+  ],
+  serviceType: 'dbass',
```
```suggestion
  serviceType: 'dbaas',
```
(See also the comment suggesting changing the type for `serviceType` to avoid these types of mistakes)

VenkatMano replied:
changed the name

venkymano replied:
changed

jdamore-linode commented on packages/manager/src/features/CloudPulse/Utils/FilterConfig.ts:
```diff
@@ -0,0 +1,130 @@
+  string,
+  CloudPulseServiceTypeFilterMap
+> = new Map([
+  ['dbass', DBASS_CONFIG],
```
```suggestion
  ['dbaas', DBAAS_CONFIG],
```

VenkatMano replied:
changed the name

jdamore-linode commented on packages/manager/src/features/CloudPulse/shared/CloudPulseComponentRenderer.test.tsx:
```diff
@@ -0,0 +1,79 @@
+    );
+
+    expect(getByPlaceholderText('Select a Region')).toBeDefined();
+  }),
+    it('it should render provided resource filter in props', () => {
+      const resourceProps = linodeFilterConfig?.filters.find(
+        (filter) => filter.configuration.filterKey == 'resource_id'
+      );
+
+      const mockDashboard = {
+        created: new Date().toDateString(),
+        id: 1,
+        label: DASHBOARD,
+        service_type: 'linode',
+        time_duration: { unit: 'min', value: 30 },
+        updated: new Date().toDateString(),
+        widgets: [],
+      };
+
+      const { getByPlaceholderText } = renderWithTheme(
+        <Grid item key={'resources'} sx={{ marginLeft: 2 }} xs>
+          {RenderComponent({
+            componentKey: 'resource_id',
+            componentProps: {
+              ...getResourcesProperties(
+                resourceProps!,
+                vi.fn(),
+                mockDashboard,
+                false,
+                { region: 'us-east' }
+              ),
+            },
+          })}
+        </Grid>
+      );
+
+      expect(getByPlaceholderText('Select Resources')).toBeDefined();
+    });
```
```suggestion
  });
  
  it('it should render provided resource filter in props', () => {
    const resourceProps = linodeFilterConfig?.filters.find(
      (filter) => filter.configuration.filterKey == 'resource_id'
    );

    const mockDashboard = {
      created: new Date().toDateString(),
      id: 1,
      label: DASHBOARD,
      service_type: 'linode',
      time_duration: { unit: 'min', value: 30 },
      updated: new Date().toDateString(),
      widgets: [],
    };

    const { getByPlaceholderText } = renderWithTheme(
      <Grid item key={'resources'} sx={{ marginLeft: 2 }} xs>
        {RenderComponent({
          componentKey: 'resource_id',
          componentProps: {
            ...getResourcesProperties(
              resourceProps!,
              vi.fn(),
              mockDashboard,
              false,
              { region: 'us-east' }
            ),
          },
        })}
      </Grid>
    );

    expect(getByPlaceholderText('Select Resources')).toBeDefined();
  });
```

VenkatMano replied:
Changed

VenkatMano commented on packages/manager/src/features/CloudPulse/Utils/FilterBuilder.ts:
```diff
@@ -0,0 +1,136 @@
+  return { '+and': filters };
+};
+
+export const checkIfWeNeedToDisableFilterByFilterKey = (
+  filterKey: string,
+  dependentFilters: {
+    [key: string]:
+      | TimeDuration
+      | number
+      | number[]
+      | string
+      | string[]
+      | undefined;
+  },
+  dashboard: Dashboard
+) => {
+  if (dashboard && dashboard.service_type) {
+    const serviceTypeConfig = FILTER_CONFIG.get(dashboard.service_type!);
+    const filters = serviceTypeConfig?.filters ?? [];
+
+    const filter = filters.find(
+      (filter) =>
+        filter?.configuration.filterKey === filterKey &&
+        filter.configuration.dependency
+    );
+
+    if (filter) {
+      return filter.configuration.dependency?.some((dependent) => {
+        const dependentFilter = dependentFilters[dependent];
+        return (
+          !dependentFilter ||
+          (Array.isArray(dependentFilter) && dependentFilter.length === 0)
+        );
+      });
+    }
+  }
+  return false;
+};
```
@jdamore-linode , the UT's should be covered as a part of component level UT's, so we didn't explicitly added here

venkymano replied:
The util functions must be covered by component level UT's

jdamore-linode replied:
That doesn't seem sufficient -- can you explain how the tests for `CloudPulseDashboardFilterBuilder` and `CloudPulseComponentRenderer` confirms that `checkIfWeNeedToDisableFilterByFilterKey` behaves correctly in case I'm missing something?

We strive to unit test our utilities and I'd be happy to share numerous examples if it'd be of any assistance.

venkymano replied:
@jdamore-linode , I added the UT's for the filterbuilder and also documented the util

VenkatMano commented on packages/manager/src/features/CloudPulse/Utils/models.ts:
```diff
@@ -0,0 +1,38 @@
+export interface CloudPulseServiceTypeFilterMap {
+  filters: CloudPulseServiceTypeFilters[];
+  serviceType: string;
```
As of now, we have linode and dbaas, and it will keep growing, so went for a string type

jdamore-linode replied:
If we need to update the code each time a new type is added anyway (e.g. `FILTER_CONFIG` const in `FilterConfig.ts`) then updating this type each time as well is trivial. Using a string literal type will catch typos that could cause real bugs for our customers.

venkymano replied:
Changed to supported linode | dbaas

VenkatMano commented on packages/manager/.changeset/pr-10718-upcoming-features-1721988074945.md:
```diff
@@ -0,0 +1,5 @@
+"@linode/manager": Upcoming Features
+---
+
+The CloudPulseDashboardFilterBuilder component builds the required mandatory global filters from a JSON config based on service types like linode, dbass etc., ([#10718](https://github.com/linode/manager/pull/10718))
```
changed

carrillo-erik commented on packages/manager/src/features/CloudPulse/shared/CloudPulseDashboardFilterBuilder.tsx:
```diff
@@ -0,0 +1,263 @@
+  REGION,
+  RELATIVE_TIME_DURATION,
+  RESOURCE_ID,
+  RESOURCES,
```
This values is being imported but not being consumed, can we please remove it.

venkymano replied:
removed

carrillo-erik commented on packages/manager/src/features/CloudPulse/shared/CloudPulseComponentRenderer.tsx:
```diff
@@ -0,0 +1,43 @@
+
+import type { CloudPulseRegionSelectProps } from './CloudPulseRegionSelect';
+import type { CloudPulseResourcesSelectProps } from './CloudPulseResourcesSelect';
+import type { Props } from './CloudPulseTimeRangeSelect';
```
Looks like you're trying to import this type but you did not export it in the corresponding file.

venkymano replied:
missed to commit that file, pushed the export changes

abailly-akamai replied:
It can't be just named `Props` - name your interfaces/type declarations better when exported so they are easily recognizable.

venkymano commented on packages/manager/src/features/CloudPulse/Utils/FilterConfig.ts:
```diff
@@ -0,0 +1,130 @@
+  serviceType: 'linode',
+};
+
+export const DBASS_CONFIG: CloudPulseServiceTypeFilterMap = {
```
done

carrillo-erik reviewed (changes requested):
The code looks much better, we appreciate you making the effort to meet the project's coding standards. The UI does require some attention, in particular the tablet and mobile views do not scale appropriately. Even without comparing to the mock ups, one can easily notice visual regressions. For example, each of the select components are different widths and the `Filters` dropdown is floating right for no apparent reason. Please refer to images below:

**Mobile View:**
![cp-mobile-view](https://github.com/user-attachments/assets/5804f741-2e19-435e-a27b-ad4c55f28858)

**Tablet View:**
![cp-tablet-view](https://github.com/user-attachments/assets/8ed3c9ed-dd5d-4880-9927-97f2cc8500b2)

**Desktop View:**
![cp-desktop-view](https://github.com/user-attachments/assets/bb279c3b-03d0-4221-955e-31a973d3578a)

venkymano commented:
Hi @carrillo-erik ,

This is just a test and functionality integration of the filterbuilder component in GlobalFilters.tsx, we have changes in our mock ups , that we we will kill some of the existing filters itself in the GlobalFilters.tsx component and fully drive the filters through our newly designed filter builder component like below screenshot

![image](https://github.com/user-attachments/assets/2bf42e6d-076b-4d3a-bf89-8ee2d1471ad5)

Once those changes comes in the upcoming PR's, the style issues will be corrected along with our updated mock up styles and we will be adding respective screenshots for mobile, tablet and desktop views. Hope this solves the case

venkymano commented:
Hi @carrillo-erik , Can this be merged now?, since we will create additional PR with the changes requested as per the above attached screenshot (latest mockups)

jdamore-linode commented:
@venkymano Can you follow up with us in Slack with a link to the ticket whose PR will contain these UI fixes?

The issues @carrillo-erik pointed out are pretty severe so we'd just like to ensure that the requirements to address this are captured and won't be backlogged since this really isn't in a state to release to customers as-is.

venkymano commented:
@jdamore-linode , https://track.akamai.com/jira/browse/DI-19816 -  As stated in the previous comments, we already have the CSS fix ready in our local integrated branch according to our latest mockups, but if we introduce that this PR will become a big PR, hence broken down the functionalities and created a small PR.

abailly-akamai commented on packages/manager/src/features/CloudPulse/shared/CloudPulseComponentRenderer.tsx:
```diff
@@ -0,0 +1,43 @@
+    });
+  }
+
+  return <Box></Box>;
```
?

venkymano replied:
We can return null in this case, I have done the changes now

abailly-akamai commented on packages/manager/src/features/CloudPulse/shared/CloudPulseDashboardFilterBuilder.tsx:
```diff
@@ -0,0 +1,262 @@
+
+    const getProps = React.useCallback(
+      (config: CloudPulseServiceTypeFilters) => {
+        if (config.configuration.filterKey == REGION) {
```
Please always use strict equality

venkymano replied:
done

abailly-akamai replied:
you need to check through the whole PR when pointed at a code consistency issue, not just what the reviewer points at

abailly-akamai commented on packages/manager/src/features/CloudPulse/shared/CloudPulseDashboardFilterBuilder.tsx:
```diff
@@ -0,0 +1,262 @@
+      return (
+        <ErrorState
+          CustomIcon={InfoIcon}
+          CustomIconStyles={{ height: '10%', width: '10%' }}
```
I don't think you'd ever want to set width/height in percentages

venkymano replied:
removed percentages

abailly-akamai commented on packages/manager/src/features/CloudPulse/shared/CloudPulseDashboardFilterBuilder.tsx:
```diff
@@ -0,0 +1,262 @@
+        <ErrorState
+          CustomIcon={InfoIcon}
+          CustomIconStyles={{ height: '10%', width: '10%' }}
+          errorText={'Please pass valid dashboard to render the filters'}
```
Do you have copy validation on those messages?

venkymano replied:
We have written UT's to validate the error message

abailly-akamai replied:
I meant copy validation from the product team

abailly-akamai commented on packages/manager/src/features/CloudPulse/shared/CloudPulseResourcesSelect.tsx:
```diff
@@ -59,7 +65,7 @@ export const CloudPulseResourcesSelect = React.memo(
       }
       // eslint-disable-next-line react-hooks/exhaustive-deps
-    }, [resources, props.region, props.resourceType]);
+    }, [resources, props.region, props.resourceType, props.xFilter]);
```
destructure your props

venkymano replied:
done

abailly-akamai commented on packages/manager/src/features/CloudPulse/Utils/FilterBuilder.ts:
```diff
@@ -0,0 +1,173 @@
+    selectedDashboard: dashboard,
+  };
+};
+
```
If you have more than 2 arguments in your function, please do not make them ordered. Use named arguments and write interfaces for their types, which can be shared within those utils.

venkymano replied:
@abailly-akamai , I have created a interface called CloudPulseFilterProperties that has common properties and used in the functions and kept a max of two arguments

abailly-akamai reviewed (changes requested):
I only scratch the surface but this PR needs a lot of work to be able to be merged.

venkymano commented on packages/manager/src/features/CloudPulse/Overview/GlobalFilters.tsx:
```diff
@@ -122,6 +124,11 @@ export const GlobalFilters = React.memo((props: GlobalFilterProperties) => {
         </Grid>
       </StyledGrid>
+      <CloudPulseDashboardFilterBuilder // this simple integration is for test purpose, the functionality integration will be done in upcoming PR by driving all filters through filterbuilder
+        dashboard={selectedDashboard!}
```
@abailly-akamai , Some missing thing in the comments, This integration remains the same, in the upcoming changes as well, we will remove the existing filters like Region and Resource selection from global filters and drive the full integration, whereas this remains the same..

venkymano replied:
@abailly-akamai , we have fixed and completed the proper integration of builder filter now, please check

venkymano commented on packages/manager/src/features/CloudPulse/Utils/FilterBuilder.ts:
```diff
@@ -0,0 +1,173 @@
+    placeholder: config.configuration.placeholder,
+    resourceType: dashboard.service_type,
+    savePreferences: !isServiceAnalyticsIntegration,
+    xFilter: buildXFilter(config, dependentFilters),
```
destructed config and used

venkymano commented on packages/manager/src/features/CloudPulse/Utils/FilterConfig.ts:
```diff
@@ -0,0 +1,132 @@
+    },
+  ],
+  serviceType: 'dbaas',
+};
```
made the serviceType and filters readonly in the model.ts, since we will use this only for configuration purposes

venkymano replied:
done

abailly-akamai replied:
quite partially...

venkymano commented on packages/manager/src/features/CloudPulse/Utils/models.ts:
```diff
@@ -0,0 +1,77 @@
+ * @placeholder - This is an optional field, controls the placeholder we need to show in autocomplete component
+ * @priority - This controls the order of the componenent, we have not used this for now, will be using in upcoming PR's
+ * @type - default is predefined filters like (region, resources, timeduration), customfilters will need to specify either static or dynamic (customselect will be implemented in upcoming PR's)
+ */
```
made the comments like JSdocs comments

venkymano commented on packages/manager/src/features/CloudPulse/Utils/FilterBuilder.ts:
```diff
@@ -0,0 +1,173 @@
+    handleStatsChange: handleTimeRangeChange,
+    key: config.configuration.filterKey,
+    placeholder: config.configuration.placeholder,
+    savePreferences: !isServiceAnalyticsIntegration,
```
@abailly-akamai , I have created a interface called CloudPulseFilterProperties and used in the functions and kept a max of two arguments

venkymano commented:
> The code looks much better, we appreciate you making the effort to meet the project's coding standards. The UI does require some attention, in particular the tablet and mobile views do not scale appropriately. Even without comparing to the mock ups, one can easily notice visual regressions. For example, each of the select components are different widths and the `Filters` dropdown is floating right for no apparent reason. Please refer to images below:
> 
> **Mobile View:** ![cp-mobile-view](https://private-user-images.githubusercontent.com/119514965/354725074-5804f741-2e19-435e-a27b-ad4c55f28858.jpg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjMwMzc0NTEsIm5iZiI6MTcyMzAzNzE1MSwicGF0aCI6Ii8xMTk1MTQ5NjUvMzU0NzI1MDc0LTU4MDRmNzQxLTJlMTktNDM1ZS1hMjdiLWFkNGM1NWYyODg1OC5qcGc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwODA3JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDgwN1QxMzI1NTFaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1jNTFkNDJiMmIwMGJmNGY0MjgwZTE4ZjA2MDEzZGMyOGRkNjFjNDUzODM2ZmU4OWRjZjYwNGMzMDk5YjYzNzUxJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.vLa-yLy09hakQALxNz0fdxrdXnGpsQGYxuIGddlP40s)
> 
> **Tablet View:** ![cp-tablet-view](https://private-user-images.githubusercontent.com/119514965/354725051-8ed3c9ed-dd5d-4880-9927-97f2cc8500b2.jpg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjMwMzc0NTEsIm5iZiI6MTcyMzAzNzE1MSwicGF0aCI6Ii8xMTk1MTQ5NjUvMzU0NzI1MDUxLThlZDNjOWVkLWRkNWQtNDg4MC05OTI3LTk3ZjJjYzg1MDBiMi5qcGc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwODA3JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDgwN1QxMzI1NTFaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0wNjcxMjQyNDlhOGIxYzRjM2E5YjYxMzhjZGExNzdiNTE1NDIxNzY3ZWVkYTM3ZjJhZmUzNjYyNzdlZWI0NTJhJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.VJ_OYww8dhh6OdsxLGtTnwftJOQRPrhHc77fX8PFvjU)
> 
> **Desktop View:** ![cp-desktop-view](https://private-user-images.githubusercontent.com/119514965/354725008-bb279c3b-03d0-4221-955e-31a973d3578a.jpg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjMwMzc0NTEsIm5iZiI6MTcyMzAzNzE1MSwicGF0aCI6Ii8xMTk1MTQ5NjUvMzU0NzI1MDA4LWJiMjc5YzNiLTAzZDAtNDIyMS05NTVlLTMxYTk3M2QzNTc4YS5qcGc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwODA3JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDgwN1QxMzI1NTFaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1jMmYwNGY4NTM2ZGQ4ZDAxOWY3YjhlYTllNjE1NjU0ZmJiZTJlNDRjNzExNmEwMzBlZjg3Zjk3YTY4ZmZkNTQzJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.xftVUQScsEWzIE2ojGF_VGZ16z_g7MGT492rkxO9Q2s)

@carrillo-erik , we have pushed the changes which had fixed this, I have added the screenshots for different devices at the top

nikhagra commented on packages/manager/src/features/CloudPulse/Overview/GlobalFilters.tsx:
```diff
@@ -1,20 +1,26 @@
+  handleAnyFilterChange(
+    filterKey: string,
+    filterValue: FilterValueType
+  ): undefined | void;
```
```suggestion
  ):  void;
```
undefined is not required as it is not going to return anything

venkymano replied:
done

nikhagra commented on packages/manager/src/features/CloudPulse/Overview/GlobalFilters.tsx:
```diff
@@ -1,20 +1,26 @@
+    filterKey: string,
+    filterValue: FilterValueType
+  ): undefined | void;
+  handleDashboardChange(dashboard: Dashboard | undefined): undefined | void;
+  handleTimeDurationChange(timeDuration: TimeDuration): undefined | void;
```
```suggestion
  handleDashboardChange(dashboard: Dashboard | undefined): void;
  handleTimeDurationChange(timeDuration: TimeDuration): void;
```

venkymano replied:
done

nikhagra commented on packages/manager/src/features/CloudPulse/Utils/FilterBuilder.ts:
```diff
@@ -0,0 +1,203 @@
+  },
+  dashboard: Dashboard
+): boolean | undefined => {
+  if (dashboard && dashboard.service_type) {
```
```suggestion
  if (dashboard?.service_type) {
```

venkymano replied:
done

abailly-akamai commented on packages/manager/src/features/CloudPulse/Utils/models.ts:
```diff
@@ -0,0 +1,182 @@
+  /**
+   * This is an optional field, it is required if the filter API response doesn't have id and we need to map it from the response
+   *
+   * @type {string}
```
You don't really need to add the JSDoc `@type` here since we are in typescript. Those are redundant

venkymano replied:
This seems redundant, I have removed those

abailly-akamai commented on packages/manager/src/features/CloudPulse/shared/CloudPulseComponentRenderer.test.tsx:
```diff
@@ -0,0 +1,87 @@
+      time_duration: { unit: 'min', value: 30 },
+      updated: new Date().toDateString(),
+      widgets: [],
+    };
```
Is there no factory for a dashboard? There should be one. This isn't typed at all.

venkymano replied:
Thanks, I created a simple dashboard factory and reused it in the test files in this PR, we will enhance the factory based on requirements moving forward

abailly-akamai commented on packages/manager/src/features/CloudPulse/shared/CloudPulseResourcesSelect.tsx:
```diff
@@ -9,27 +10,42 @@ import {
   handleResourcesSelection: (resources: CloudPulseResources[]) => void;
   placeholder?: string;
-  region: string | undefined;
+  region?: string | undefined;
```
either 

`region: string | undefined;`
or 
`region?: string`

not both

venkymano replied:
This is really great suggestion, i have made use region?: string

abailly-akamai commented on packages/manager/src/features/CloudPulse/shared/CloudPulseResourcesSelect.tsx:
```diff
@@ -9,27 +10,42 @@ import {
     const { data: resources, isLoading } = useResourcesQuery(
-      props.region && props.resourceType ? true : false,
-      props.resourceType,
+      disabled != undefined ? !disabled : Boolean(region && resourceType),
```
strict equality please

venkymano replied:
done, checked elsewhere as well for the files changed in the PR

abailly-akamai commented on packages/manager/src/features/CloudPulse/shared/CloudPulseResourcesSelect.tsx:
```diff
@@ -9,27 +10,42 @@ import {
   label: string;
+  placeholder?: string;
   region?: string; // usually linodes are associated with only one region
   regions?: string[]; // aclb are associated with multiple regions
```
This is just strange. why not handling just one required prop?? You can just pass a single region in an array. This just complicates code.

venkymano replied:
The instances fetch API of linode returns region where as instances fetch API of aclb returns regions, to achieve this we need to introduce intercept logic that will take care of converting to one region, that is also a complex thing, so we are maintaining it like this.

abailly-akamai replied:
This does not make sense to me

nikhagra replied:
@abailly-akamai could you please suggest an alternative to it

abailly-akamai replied:
@nikhagra for one, ACLB has been sunsetted, so I doubt you need to support it. 

And if you really need to support multiple regions, then it should be the only prop.

nikhagra replied:
@abailly-akamai  I understand, but as we are building a configuration based filter & it is not limited only to ACLB or linodes but any service type could be integrated to it and we don't have confirmed information whether all the service types ( existing or will come in future ) that'll be integrated to it will have single region or multiple regions

jaalah-akamai replied:
Not to prolong this discussion but an alternative could be something like `regionOrRegions?: string | string[];` and then do some type narrowing where needed. Just something to think about!

venkymano replied:
@jaalah / @abailly-akamai , since we don't need the regions prop as of now and we have not used in this component, we will enhance the component when we integrate with the service where this use case comes

abailly-akamai commented on packages/manager/src/features/CloudPulse/shared/CloudPulseResourcesSelect.tsx:
```diff
@@ -70,20 +86,32 @@ export const CloudPulseResourcesSelect = React.memo(
+    oldProps.region === newProps.region &&
+    oldProps.resourceType === newProps.resourceType &&
+    deepEqual(oldProps.xFilter, newProps.xFilter)
+  );
```
this is a bit smelly. may be trouble in the long term.

You should optimize this to future proof implementation

Pseudo/Kinda functional code

```Typescript
const compareProps = (
  prevProps: CloudPulseResourcesSelectProps,
  nextProps: CloudPulseResourcesSelectProps
): boolean => {
  const keysToCompare: (keyof CloudPulseResourcesSelectProps)[] = [
    'region',
    'resourceType',
    'disabled',
    'placeholder',
    'savePreferences'
  ];

  for (const key of keysToCompare) {
    if (prevProps[key] !== nextProps[key]) {
      return false;
    }
  }

  // Deep comparison for xFilter
  if (!deepEqual(prevProps.xFilter, nextProps.xFilter)) {
    return false;
  }

  // Ignore function props in comparison
  return true;
};
```

It's easier to add or remove properties to compare as the component evolves. It short-circuits on the first difference found, potentially saving unnecessary comparisons.
Also, using `keyof CloudPulseResourcesSelectProps` ensures type safety if prop names change and still uses deepEqual for the xFilter prop, which might be an object.

Test it well, tho, it may need tweaking

venkymano replied:
Since we compare only region and resourceType that is required for re rendering, I updated the code with the suggested function for props region, resourceType and xFilter alone and it works fine

abailly-akamai commented on packages/manager/src/features/CloudPulse/shared/CloudPulseTimeRangeSelect.tsx:
```diff
@@ -34,52 +36,48 @@ export type Labels =
+        disableClearable
+        fullWidth
+        isOptionEqualToValue={(option, value) => option.value === value.value}
+        label=""
```
This is not accessible without a label or aria equivalent

abailly-akamai replied:
@venkymano you'd think it's required for a reason 😄 ? Passing an empty string is breaking accessibility and going around the problem. You have to take in consideration users using assistive technology and an input without a label makes identifying an input quite hard to for a screen reader for instance.

In this case, (and all other instances of Autocomplete in CloudPulse), pass a proper label and add `hideLabel` to its `inputProps` to visually hide it but keep it in the markup.

jaalah-akamai replied:
Should be all set.

abailly-akamai reviewed (changes requested):
Thanks for addressing comments and improving best practices. In general the code quality is still pretty rough for having been reviewed several times by your team and ours. Notably, when pointing at a code issue, I would expect the author to check the PR for similar defects, not just addressed the ones we pointed at. If not, the back and forth 
takes too much time out of everyone. There's frankly a lot more to pick at than I reviewed, but tried to only address the biggest concerns.

Apart from code review comments, please address the issues below.

You're updating state during re-render, which points to issues in your lifecycle methods.  This needs to be addressed before merging

![Screenshot 2024-08-08 at 09 32 20](https://github.com/user-attachments/assets/d4f74e72-9f8a-4cd9-8ef0-afb2cb6e4751)

---

UI looks improved minus a few issues:
![Screenshot 2024-08-08 at 09 43 09](https://github.com/user-attachments/assets/4743f223-a2ba-4885-a4d9-d962b3728f86)

![Screenshot 2024-08-08 at 10 25 22](https://github.com/user-attachments/assets/d4d1fe7e-4fd7-4806-b609-f78e4cda6daf)

![Screenshot 2024-08-08 at 10 26 27](https://github.com/user-attachments/assets/8c9ba275-24a0-4d4e-b579-9075e2d5ac54)

![Screenshot 2024-08-08 at 10 36 38](https://github.com/user-attachments/assets/7c9d04ef-883b-421e-be91-eb50a40cc9d0)

These rendering flickers are problematic and point to issues with the code, perhaps when updating preferences

https://github.com/user-attachments/assets/565f0566-3fad-464e-9d12-5d3e835e431f

abailly-akamai commented:
@venkymano considering the quality of the PR, and looking at your changelog, we need smaller PRs in the future and big PRs will be asked to be broken down:

- Added a new CloudPulseDashboardFilterBuilder that will build the global filters based on FilterConfig.ts per service type
- Added a new CloudPulseComponentRenderer component that builds the react element based on filterKey
- Complete integration of the CloudPulseDashboardFilterBuilder in the GlobalFilters.tsx
- Added logic in the CloudPulseDashboardLanding component to pass the filters selected to the dashboard component
- Changes for moving CloudPulseTimeRangeSelect from select to Autocomplete
- CSS changes making the filters compatible with different screens like desktop, tablet and mobile.

almost all of those changes should be its own PR.

venkymano commented on packages/manager/src/features/CloudPulse/Dashboard/CloudPulseDashboardLanding.tsx:
```diff
@@ -1,37 +1,137 @@
+      setTimeDuration(timeDurationObj);
+    },
+    []
+  );
```
Thanks for the suggestion, Since we update the preference on every filter change and the preference update will make the cloudpulsedashboard to re-render, it is better to use with callback functionality itself

venkymano commented on packages/manager/src/features/CloudPulse/Utils/FilterBuilder.test.ts:
```diff
@@ -0,0 +1,181 @@
+
+  const result = getResourcesProperties(
+    {
+      config: resourceSelectionConfig!,
```
@abailly-akamai , Thanks for the suggestion, if you see the line number 90 we check

`expect(resourceSelectionConfig).toBeDefined();`

So if the resourceSelectionConfig is really undefined , it will not even reach the below lines and test case fails, since we made sure that resourceSelectionConfig is never undefined here, added a non-nullish assertion. Please Let me know if we still need this to be handled

venkymano commented on packages/manager/src/features/CloudPulse/Dashboard/CloudPulseDashboardLanding.tsx:
```diff
@@ -1,37 +1,137 @@
+            : []
+        }
+        dashboardId={dashboard!.id!}
+        duration={timeDuration!}
```
Thanks, we don't need for dashboardId, but if you see the if condition above the return statement, we are already checking for timeDuration is defined or not in the function checkIfAllMandatoryFiltersAreSelected, if this is false, we will not reach the statement, so we are sure here timeDuration is not undefined, but I have added one more check and removed these.

jdamore-linode replied:
@venkymano That may be correct for the code in its current state, but using the non-nullish assertion here creates a maintenance risk where changes to the code (e.g. changing or removing `checkIfAllMandatoryFiltersAreSelected`) could result in `timeDuration` being `null`, and TypeScript won't catch it thanks to the non-nullish assertion.

Things like non-nullish assertions and `any` should be treated as escape hatches to circumvent TypeScript's safety checks when it is absolutely necessary _only_. I don't think any of the non-nullish assertions that @abailly-akamai pointed out meet that bar.

venkymano replied:
@jdamore-linode , i have fixed these now, removed nonnullish assertions

venkymano commented:
@abailly-akamai 

Here are some style issue fixes according to the screenshot we made

1. Fixed the console issue
2. Fixed the alignment issue of manual refresh button
3. Removed the border in the globalfilters, since we don't need it.
4. Made the error clear after selecting some values

abailly-akamai commented:
@venkymano the rendering flickers as shown here and pointed at in the previous review need to be fixed as well

https://github.com/user-attachments/assets/002341e3-68c0-4076-a395-abdc8d19226e

jaalah-akamai commented:
@abailly-akamai I made a few of the changes you requested. The flickering is a comment I mentioned [in previous PRs](https://github.com/linode/manager/pull/10710#issuecomment-2255879162) too. They have a ticket to address this when the endpoints are in. Anything else that needs to be looked at?

nikhagra commented:
> @abailly-akamai I made a few of the changes you requested. The flickering is a comment I mentioned [in previous PRs](https://github.com/linode/manager/pull/10710#issuecomment-2255879162) too. They have a ticket to address this when the endpoints are in. Anything else that needs to be looked at?

Thanks @jaalah to fix few styling. Here is the ticket to handle flickering issue. Will take care of it.

https://track.akamai.com/jira/browse/DI-20169

jaalah-akamai commented:
@nikhagra Let's resolve the conflict so we can kick off another CI run in preparation to merge 👍

venkymano commented:
@jaalah-akamai , I have resolved the conflicts now

carrillo-erik commented on packages/manager/src/features/CloudPulse/shared/CloudPulseComponentRenderer.tsx:
```diff
@@ -0,0 +1,47 @@
+  resource_id: CloudPulseResourcesSelect,
+};
+
+const buildComponent = (props: CloudPulseComponentRendererProps) => {
```
In the future, remember to use PascalCase for component names and aim for named exports over default exports.
